Exhaustive match checking doesn't work on strings

## Reproduction steps

Scala version: 2.13.10

compile the following with `scalac -Xlint Test.scala`
```scala
object Test {
  def doesntWarn(input: String) = input match {
    case "a" =>
    case "b" =>
  }
  def warns(input: String) = input match {
    case "b" =>
  }
}
```
The output is
```
Test.scala:8: warning: match may not be exhaustive.
It would fail on the following input: (x: String forSome x not in "b")
  def warns(input: String) = input match {
                             ^
1 warning
```

## Problem

There are two inexhaustive matches, there should be two warnings

Both `match` statements are clearly inexhaustive, but match statements on `String`s (or `Int`s or `Char`s and probably other things too) with two or more literal cases don't trigger a `match not may be exhaustive` warning.

Adding a third `case "c" if false =>` to `doesntWarn` results in an inexhaustive match warning and an unreachable code warning.

Adding just `case "c" =>` or `case other if other.isEmpty =>` (any condition will do as long as it's not a boolean literal, afaict) has no effect.

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala
@@ -640,13 +640,36 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
               case a :: b :: c :: _ => 3
               case cases            => cases.map {
                 case AlternativesTreeMaker(_, alts, _) :: _ => lengthMax3(alts)
-                case c                                      => 1
+                case _                                      => 1
               }.sum
             }
             lengthMax3(cases) > 2
           }
           hasSwitchAnnotation && exceedsTwoCasesOrAlts
         case _ => false
+      }
+    }
+
+    // See the use of RegularSwitchMaker by SwitchEmission#emitSwitch, which this code emulates or duplicates.
+    private object Switchable {
+      val switchableTpe = Set(ByteTpe, ShortTpe, IntTpe, CharTpe, StringTpe)
+
+      def apply(scrutSym: Symbol, cases: List[List[TreeMaker]]): Boolean = switchableTpe(scrutSym.tpe.dealiasWiden) && {
+        def switchable(tms: List[TreeMaker]): Boolean =
+          tms.forall {
+            case EqualityTestTreeMaker(_, SwitchablePattern(), _) => true
+            case AlternativesTreeMaker(_, altss, _) => Switchable(scrutSym, altss)
+            case BodyTreeMaker(_, _) => true
+            case _ => false
+          }
+        cases.forall(switchable)
+      }
+
+      object SwitchablePattern {
+        def unapply(pat: Tree): Boolean = pat.tpe match {
+          case const: ConstantType => const.value.isIntRange || const.value.tag == StringTag || const.value.tag == NullTag
+          case _ => false
+        }
       }
     }
 
@@ -665,6 +688,12 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
           if (requiresSwitch(scrut, cases))
             typer.context.warning(scrut.pos, "could not emit switch for @switch annotated match", WarningCategory.OtherMatchAnalysis)
 
+          // If cases are switchable, suppress warning for exhaustivity.
+          // The switch was not emitted, probably because there aren't enough cases.
+          val suppression1 =
+            if (Switchable(scrutSym, cases)) suppression.copy(suppressExhaustive = true)
+            else suppression
+
           if (!cases.isEmpty) {
             // before optimizing, check cases for presence of a default case,
             // since DCE will eliminate trivial cases like `case _ =>`, even if they're the last one
@@ -677,7 +706,7 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
               case _                              => matchFailGen
             }
 
-            analyzeCases(scrutSym, cases, pt, suppression)
+            analyzeCases(scrutSym, cases, pt, suppression1)
 
             val (optimizedCases, toHoist) = optimizeCases(scrutSym, cases, pt, selectorPos)
 

--- a/test/files/neg/t12770.check
+++ b/test/files/neg/t12770.check
@@ -1,0 +1,11 @@
+t12770.scala:12: warning: match may not be exhaustive.
+It would fail on the following input: B()
+  def control(input: T) = input match {
+                          ^
+t12770.scala:15: warning: match may not be exhaustive.
+It would fail on the following input: (x: Any forSome x not in ("a", "b", "c", 42))
+  def any(input: Any) = input match {
+                        ^
+error: No warnings can be incurred under -Werror.
+2 warnings
+1 error

--- a/test/files/neg/t12770.scala
+++ b/test/files/neg/t12770.scala
@@ -1,0 +1,25 @@
+
+// scalac: -Werror -Xlint
+
+object Test {
+  def doesntWarn(input: String) = input match {
+    case "a" =>
+    case "b" =>
+  }
+  def warnsNoLonger(input: String) = input match {
+    case "c" =>
+  }
+  def control(input: T) = input match {
+    case _: A =>
+  }
+  def any(input: Any) = input match {
+    case "a" =>
+    case 42  =>
+    case "b" =>
+    case "c" =>
+  }
+}
+
+sealed trait T
+final class A extends T
+final class B extends T

